plugins/woocommerce/tests/php/src/Utilities/CallbackUtilTest.php

```diff
@@ -0,0 +1,377 @@
+		$object    = new DummyCallbackClass();
+		$callback1 = array( $object, 'my_method' );
+		$callback2 = array( $object, 'another_method' );
+
+		$signature1 = CallbackUtil::get_callback_signature( $callback1 );
+		$signature2 = CallbackUtil::get_callback_signature( $callback2 );
+
+		$this->assertNotEquals( $signature1, $signature2 );
+
+		$this->assertEquals(
+			DummyCallbackClass::class . '::my_method',
+			$signature1
+		);
+
+		$this->assertEquals(
+			DummyCallbackClass::class . '::another_method',
+			$signature2
+		);
+	}
+
+	/**
+	 * @testdox `get_callback_signature` should produce different signatures for the same method name on different classes.
+	 */
+	public function test_get_callback_signature_same_method_different_classes() {
+		$object1   = new DummyCallbackClass();
+		$object2   = new AnotherDummyCallbackClass();
+		$callback1 = array( $object1, 'my_method' );
+		$callback2 = array( $object2, 'my_method' );
+
+		$signature1 = CallbackUtil::get_callback_signature( $callback1 );
+		$signature2 = CallbackUtil::get_callback_signature( $callback2 );
+
+		$this->assertNotEquals( $signature1, $signature2 );
+
+		$this->assertEquals(
+			DummyCallbackClass::class . '::my_method',
+			$signature1
+		);
+
+		$this->assertEquals(
+			AnotherDummyCallbackClass::class . '::my_method',
+			$signature2
+		);
+	}
+
+	/**
+	 * @testdox `get_callback_signature` should fall back to serialization for invalid array callbacks with non-string method.
+	 */
+	public function test_get_callback_signature_with_invalid_array_non_string_method() {
+		$object   = new DummyCallbackClass();
+		$callback = array( $object, 123 );
+
+		$signature = CallbackUtil::get_callback_signature( $callback );
+
+		// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
+		$this->assertEquals( serialize( $callback ), $signature );
+	}
+
+	/**
+	 * @testdox `get_callback_signature` should fall back to serialization for invalid array callbacks with non-object/non-string target.
+	 */
+	public function test_get_callback_signature_with_invalid_array_non_object_target() {
+		$callback = array( 123, 'method' );
+
+		$signature = CallbackUtil::get_callback_signature( $callback );
+
+		// phpcs:ignore WordPress.PHP.DiscouragedPHPFunctions.serialize_serialize
+		$this->assertEquals( serialize( $callback ), $signature );
+	}
+
+	/**
+	 * @testdox `get_hook_callback_signatures` should return empty array for non-existent hook.
+	 */
+	public function test_get_hook_callback_signatures_with_non_existent_hook() {
+		$signatures = CallbackUtil::get_hook_callback_signatures( 'non_existent_hook_12345' );
+
+		$this->assertIsArray( $signatures );
+		$this->assertEmpty( $signatures );
+	}
+
+	/**
+	 * @testdox `get_hook_callback_signatures` should return signatures for all callbacks on a hook.
+	 */
+	public function test_get_hook_callback_signatures_with_multiple_callbacks() {
+		$hook_name = 'test_hook_' . wp_rand();
```

Comment:
Any reason the hook names need to use `wp_rand()`? Not saying this is the heaviest function (though it's not simple either), but seems a bit unnecessary.